packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,241 @@
+  | 'ext_ublock'
+  | 'ext_ubol'
+  | 'ext_devbuild'
+  | 'env_chromium'
+  | 'env_edge'
+  | 'env_firefox'
+  | 'env_mobile'
+  | 'env_safari'
+  | 'env_mv3'
+  | 'false'
+  | 'cap_html_filtering'
+  | 'cap_user_stylesheet'
+  | 'adguard'
+  | 'adguard_app_android'
+  | 'adguard_app_ios'
+  | 'adguard_app_mac'
+  | 'adguard_app_windows'
+  | 'adguard_ext_android_cb'
+  | 'adguard_ext_chromium'
+  | 'adguard_ext_edge'
+  | 'adguard_ext_firefox'
+  | 'adguard_ext_opera'
+  | 'adguard_ext_safari'
+  | (string & {});
+
+export class Env extends Map<EnvKeys, boolean> {}
+
+export const enum PreprocessorTypes {
+  INVALID = 0,
+  BEGIF = 1,
+  ELSE = 2,
+  ENDIF = 3,
+}
+
+export function detectPreprocessor(line: string) {
+  if (line.charCodeAt(1) !== 35 /* '#' */) {
+    return PreprocessorTypes.INVALID;
+  }
+
+  const command = line.slice(2);
+
+  if (fastStartsWith(command, 'if ')) {
+    return PreprocessorTypes.BEGIF;
+  }
+
+  if (command === 'else') {
+    return PreprocessorTypes.ELSE;
+  }
+
+  if (command === 'endif') {
+    return PreprocessorTypes.ENDIF;
+  }
+
+  return PreprocessorTypes.INVALID;
+}
+
+const tokenizerPattern = /(!|&&|\|\||\(|\)|[a-zA-Z0-9_]+)/g;
+
+const tokenize = (expression: string) => expression.match(tokenizerPattern);
+
+const precedence: Record<string, number> = {
+  '!': 2,
+  '&&': 1,
+  '||': 0,
+};
+
+const isOperator = (token: string) => Object.prototype.hasOwnProperty.call(precedence, token);
+
+/// The parsing is done using the [Shunting yard algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
+/// This function takes as input a string expression and an environment Map.
+/// The expression is made of constants (identifiers), logical operators
+/// (&&, ||), negations (!constant) and parentheses.
+///
+/// The environment is a simple Map that associates identifiers to boolean values.
+///
+/// The function should return the result of evaluating the expression using
+/// the values from `environment`. The return value of this function is
+/// either `true` or `false`.
+export const evaluate = (expression: string, env: Env) => {
+  const length = expression.length;
+
+  // Fast path for very simple conditions (which are at the moment 100% of all
+  // preprocessor conditions: either a single constant or a negated constant)
+  if (!length) {
+    return false;
+  }
+
+  const tokens = tokenize(expression);
+
+  if (!tokens) {
+    return false;
+  }
+
+  const output: (boolean | string)[] = [];
+  const stack: (boolean | string)[] = [];
+
+  let size = 0;
+
+  for (const token of tokens) {
+    size += token.length;
+
+    if (token === '(') {
+      stack.push(token);
+    } else if (token === ')') {
+      const parenthesesOpenIndex = stack.lastIndexOf('(');
+
+      // If the opening parenthesis doesn't exist
+      if (parenthesesOpenIndex < 0) {
+        return false;
+      }
+
+      output.push(...stack.splice(parenthesesOpenIndex + 1, stack.length - parenthesesOpenIndex));
```

Comment:
I am not sure this optimization is correct. You need to revert the order because the original algorithm loops over elements from the stack backward, and pushes them to output, which should reverse the order. Let's use a simple loop for clarity.

```js
      while (stack.length !== 0 && stack[stack.length - 1] !== '(') {
        output.push(stack.pop());
      }

      // If the opening parenthesis doesn't exist
      if (stack.length === 0) {
        return false;
      }

      stack.pop();
```

Reply:
Checked, I reverted the optimisation.